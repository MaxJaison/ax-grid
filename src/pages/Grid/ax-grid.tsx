import { Table, Thead, Tbody, Tr, Th, Td, TableContainer, Stack, useDisclosure, Button } from '@chakra-ui/react';
import { Filters } from '../../components/Filters.tsx';
import { DEFAULT_COLUMNS, ENERGY_TYPES } from '../../utils/constants.ts';
import { NewDealModal } from '../../components/NewDealModal.tsx';
import { useEffect, useState } from 'react';
import { socket } from '../../utils/socket.ts';
import { getCustomColumns } from '../../utils/mocks.ts';

export const AxGrid = () => {
    const { isOpen, onOpen, onClose } = useDisclosure();
    const [allData, setAllData] = useState<string[][]>([]);
    const [energyType, setEnergyType] = useState(ENERGY_TYPES[0]);

    useEffect(() => {
        const onConnect = () => {
            console.log('connected');
        };

        const onDisconnect = () => {
            console.log('disconnected');
        };

        const onSendEnergy = (energy: string) => {
            setEnergyType(energy);
        };

        const onAllTradeEvent = (allTrades: string[][]) => {
            setAllData(allTrades);
        };

        socket.on('connect', onConnect);
        socket.on('disconnect', onDisconnect);
        socket.on('all-trades', onAllTradeEvent);
        socket.on('send-energy', onSendEnergy);

        return () => {
            socket.off('connect', onConnect);
            socket.off('disconnect', onDisconnect);
            socket.off('all-trades', onAllTradeEvent);
            socket.off('send-energy', onSendEnergy);
        };
    }, []);

    const proceedTrade = (id: number) => {
        socket.emit('trade-status', { id, status: 'Processing' });
    };

    const madeTrade = (id: number) => {
        socket.emit('trade-status', { id, status: 'Completed' });
    };

    return (
        <Stack m="20px" spacing="10">
            <Filters onOpen={onOpen} socket={socket} />
            <TableContainer overflow="auto">
                <Table variant="striped" colorScheme="teal">
                    <Thead color="#F7F9FF">
                        <Tr>
                            {DEFAULT_COLUMNS.map((column: string, index: number) => {
                                return <Th key={index}>{column}</Th>;
                            })}
                            {getCustomColumns(energyType).map((column: string, index: number) => {
                                return <Th key={index}>{column}</Th>;
                            })}
                            <Th>Action</Th>
                        </Tr>
                    </Thead>
                    <Tbody>
                        {allData
                            .filter((row) => row[2] === energyType)
                            .map((row, index: number) => {
                                return row ? (
                                    <Tr key={index}>
                                        {row.map((value: string, ind: number) => {
                                            if (ind == 0) {
                                                return;
                                            }
                                            return <Td key={ind}>{value}</Td>;
                                        })}
                                        <Td>
                                            {row[1] === 'Open' ? (
                                                <Button colorScheme="blue" onClick={() => proceedTrade(row[0])}>
                                                    Trade
                                                </Button>
                                            ) : row[1] === 'Processing' ? (
                                                <Button colorScheme="red" onClick={() => madeTrade(row[0])}>
                                                    Complete
                                                </Button>
                                            ) : (
                                                'Completed'
                                            )}
                                        </Td>
                                    </Tr>
                                ) : (
                                    <Tr></Tr>
                                );
                            })}
                    </Tbody>
                </Table>
            </TableContainer>
            <NewDealModal isOpen={isOpen} onClose={onClose} socket={socket} energyType={energyType} />
        </Stack>
    );
};
