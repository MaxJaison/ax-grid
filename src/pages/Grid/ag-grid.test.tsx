import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import { AxGrid } from './ax-grid.tsx';

describe('AxGrid Component', () => {
    it('renders the AxGrid component', async () => {
        render(<AxGrid />);

        expect(screen.getByText(/Add New Deal/i)).toBeInTheDocument();
    });

    it('create AxGrid snapshot', async () => {
        const { asFragment } = render(<AxGrid />);
        expect(asFragment()).toMatchSnapshot();
    });
});
