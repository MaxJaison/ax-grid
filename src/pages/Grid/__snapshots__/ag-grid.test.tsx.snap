// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AxGrid Component > create AxGrid snapshot 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1umn640"
  >
    <div
      class="css-1jr4x0c"
    >
      <p
        class="chakra-text css-bxx4ns"
      >
        Select which type of energy to display in the grid: 
      </p>
      <div
        class="chakra-select__wrapper css-18sl0nf"
      >
        <select
          class="chakra-select css-1j3r4nt"
        >
          <option
            selected=""
            value="Solar"
          >
            Solar
          </option>
          <option
            value="Gas"
          >
            Gas
          </option>
          <option
            value="Wind"
          >
            Wind
          </option>
          <option
            value="Hydro"
          >
            Hydro
          </option>
          <option
            value="Kinetic"
          >
            Kinetic
          </option>
          <option
            value="Thermal"
          >
            Thermal
          </option>
        </select>
        <div
          class="chakra-select__icon-wrapper css-162mkon"
        >
          <svg
            aria-hidden="true"
            class="chakra-select__icon"
            focusable="false"
            role="presentation"
            style="width: 1em; height: 1em; color: currentColor;"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <button
        class="chakra-button css-51wpq4"
        type="button"
      >
        Add New Deal
      </button>
    </div>
    <div
      class="chakra-table__container css-zipzvv"
    >
      <table
        class="chakra-table css-0"
      >
        <thead
          class="css-fesj20"
        >
          <tr
            class="css-0"
          >
            <th
              class="css-0"
            >
              Trade Status
            </th>
            <th
              class="css-0"
            >
              Energy Type
            </th>
            <th
              class="css-0"
            >
              Price
            </th>
            <th
              class="css-0"
            >
              Min Purchase Quantity
            </th>
            <th
              class="css-0"
            >
              Contract Terms
            </th>
            <th
              class="css-0"
            >
              Payment Terms
            </th>
            <th
              class="css-0"
            >
              Capacity
            </th>
            <th
              class="css-0"
            >
              Location
            </th>
            <th
              class="css-0"
            >
              Energy Output Predictions
            </th>
            <th
              class="css-0"
            >
              Time Of Availability
            </th>
            <th
              class="css-0"
            >
              Certifications
            </th>
            <th
              class="css-0"
            >
              Action
            </th>
          </tr>
        </thead>
        <tbody
          class="css-0"
        />
      </table>
    </div>
  </div>
</DocumentFragment>
`;
